### PR TITLE
Dialogs app file browser wrapper

### PR DESCRIPTION
**Pull request by mogenson**
Add a safe Rust wrapper for the dialog file browser methods and the
DialogsFileBrowserOptions type. Rename the DialogsApp::show() method
to show_message() and create new show_file_browser() method. Add a
getter method to FuriString to provide access to the inner
sys::FuriString via a raw pointer because a sys::FuriString is needed by
a few of the Furi C API functions. Extend the storage example to open a
file browser and have the user select the file that was just written to
the file system.

Note: The Furi C struct is named DialogsFileBrowserOptions, but all
other dialog types and functions are singular instead of plural. The
Rust type is named DialogFileBrowserOptions.

The DialogFileBrowserOptions type uses setter methods that return Self
so they can be chained together in the builder pattern. The
DialogFileBrowserOptions includes a Rust unit type phantom data member.
The type or size does not really matter, but PhantomData does require a
sized type. We just need one struct member to carry the lifetime
parameter so that CStr and Icon references are not dropped until the
struct goes out of scope.

### CODE REVIEW

**Review by JarvisCraft — CHANGES_REQUESTED** [review]

File: crates/flipperzero/src/dialogs/mod.rs
Comment (mogenson) [reply crates/flipperzero/src/dialogs/mod.rs:90]: Thanks for the tip! I will change.

File: crates/flipperzero/src/dialogs/mod.rs
Comment (mogenson) [reply crates/flipperzero/src/dialogs/mod.rs:209]: I was following the pattern found in C code of creating an uninitialized struct: `DialogsFileBrowserOptions = {};` and passing it to `dialog_file_browser_set_basic_options()` to initialize.

I can initialize all the struct values how `dialog_file_browser_set_basic_options()` does in Rust and avoid the call into C. The options can then be customized with the setter methods. I think I'll remove the `Default` trait implementation because `new()` will be the default.

File: crates/flipperzero/src/dialogs/mod.rs
Comment (mogenson) [reply crates/flipperzero/src/dialogs/mod.rs:221]: This shouldn't be moving the address of `dialog_file_browser_options` we are passing a raw pointer to the inner `sys::DialogsFileBrowserOptions` for the C function to populate. This is just a helper function that sets things like `options->icon = icon;` in C instead of us assigning the variable directly in Rust. Anyways, I'm removing this FFI call and now doing all the default value assignments in Rust.

Comment (JarvisCraft) [reply crates/flipperzero/src/dialogs/mod.rs]: Actually, we can use `dialog_file_browser_set_basic_options` safely (from API standpoint) which would allow us to not copy the defaults (which may be important if the deafult change in C impl) and yet keep soundness: you could use `MaybeUninit` for stack-allocating the structure and then passing it to `dialog_file_browser_set_basic_options`.

File: crates/flipperzero/src/dialogs/mod.rs
Comment (mogenson) [reply crates/flipperzero/src/dialogs/mod.rs]: Check out the current change. By the way, in the Furi C code, some pointer values are allowed to be NULL. For example, the `base_path`, `icon, and `item_loader_context` can all be NULL, these are checked in the C code. The `extension` member has to be a valid string. When calling `sys::dialog_file_browser_show()`, you can also pass NULL for the file browser options.

Comment (JarvisCraft) [reply crates/flipperzero/src/dialogs/mod.rs]: My apologize for a bit chaotic review :heart:

File: crates/flipperzero/src/dialogs/mod.rs
Comment (mogenson) [reply crates/flipperzero/src/dialogs/mod.rs]: Ah, well now that you've seen things both ways do you have a strong preference? I also just noticed that `dialog_file_browser_set_basic_options()` doesn't touch the `hide_dot_files` member. It seems to rely on a struct being zeroed and zero being a falsey value in C.

File: crates/flipperzero/src/dialogs/mod.rs
Comment (JarvisCraft) [inline crates/flipperzero/src/dialogs/mod.rs:79]: I fell like this is unsound:

1. `options` is passed by `&` and the return type has no lifetimes;
2. you convert `&opts.data` to a `const`-pointer;
3. you return from the function thus the original `&DialogFileBrowserOptions` can now be moved.

Now if the API user moves the `DialogFileBrowserOptions` to which he pointed in the constructor, it will include moving its `data` field which will invalidate the `options` pointer passed to `dialog_file_browser_show`

**Review by JarvisCraft — CHANGES_REQUESTED** [review]

File: crates/flipperzero/src/dialogs/mod.rs
Comment (JarvisCraft) [reply crates/flipperzero/src/dialogs/mod.rs]: That's a really interesting question.

It may just be secretly unsound in flipperzero-firmware since their `DialogsFileBrowserOptions browser_options` initialization does not guarantee zero-filling (AFAIK).

Let's keep rust-based initialization as for now, since it is easier to speculate on its safety as for now. But I will ask firmware maintainers if it's just an overlooked field in their initialization and we can ask them to make `dialog_file_browser_set_basic_options()` always fully initialize the instance so that we can rely on it (and in this case we will add a future PR to replace current approach with `MaybeUnit`-based).

Comment (JarvisCraft) [reply crates/flipperzero/src/dialogs/mod.rs]: I've [asked](https://t.me/flipperzero/103925/177467) them but I suggest that we don't wait fr the response (since it also requires filing a PR to firmware, a new release of firware and update of our bindings) and use Rust-approach by now.

File: crates/flipperzero/src/dialogs/mod.rs
Comment (mogenson) [reply crates/flipperzero/src/dialogs/mod.rs:79]: I think this is ok, but I'm not an expert on Rust lifetimes.
1. `options` is only used as a reference and is unmodified by this Rust function or the `dialog_file_browser_show` C function. The C function is blocking, when it returns, we are done with the `options` and it can be dropped by the caller when it goes out of scope. So I don't think we need to move `options` into `show_file_browser` or hold onto a reference.
2. Yep, `dialog_file_browser_show` wants a const pointer to a `DialogsFileBrowserOptions` (the inner C struct). The const designates that C won't modify this struct and we can inspect the C code to know that it won't hold onto this pointer after the `dialog_file_brower_show` function returns.
3. The caller can do whatever they want to the original `DialogFileBrowserOptions` after this function, no data will be modified from underneath them.

Comment (JarvisCraft) [reply crates/flipperzero/src/dialogs/mod.rs:79]: Hmm, I haven't thought that `dialog_file_browser_show` blocks until it has done all its work.

In this case, yes, everyhting seems fine :rocket:

File: crates/flipperzero/src/dialogs/mod.rs
Comment (JarvisCraft) [inline crates/flipperzero/src/dialogs/mod.rs:35]: Extra small detail: I guess `()` is a more common type to be used in `PhantomReference`:

```suggestion
    _phantom: PhantomData<&'a ()>,
```

**Review by JarvisCraft — CHANGES_REQUESTED** [review]

**Comment by mogenson** [comment]
@JarvisCraft are you ok with merging this now?

**Comment by JarvisCraft** [comment]
I actually am not the one with push permissions, so I summon @dcoles and @str4d for this purpose :)

File: crates/flipperzero/src/dialogs/mod.rs
Comment (dcoles) [inline crates/flipperzero/src/dialogs/mod.rs:74]: You may wish to consider:
```rust
let options = options.map(|opts| &opts.data as *const sys::DialogsFileBrowserOptions).unwrap_or(ptr::null());
```

File: crates/flipperzero/src/dialogs/mod.rs
Comment (dcoles) [inline crates/flipperzero/src/dialogs/mod.rs:202]: You can also use `sys::c_string!` here too.

File: crates/flipperzero/src/furi/string.rs
Comment (dcoles) [inline crates/flipperzero/src/furi/string.rs:90]: This should take `&mut self` and be called `as_mut` since this effectively allows you mutable access to the string data.

File: crates/flipperzero/examples/storage.rs
Comment (dcoles) [inline crates/flipperzero/examples/storage.rs:28]: We should use the `sys::c_string!` macro here (it automatically adds a b"\0").

**Review by dcoles — CHANGES_REQUESTED** [review]
Hi @mogenson,

Thanks for the PR! This looks great.

Most of my feedback is just minor style things, though I'd say the `as_ptr` returning a `*mut sys::FuriString` is important to address before I can merge this.

File: crates/flipperzero/src/furi/string.rs
Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string.rs:90]: I would disagree. `core` and `std` types do not require `mut` on `self` since the raw pointer itself cannot be used in any bad way, while its dereferenceing is anyway `unsafe`.

Also our existing APIs use `&self`, for example `UnsafeRecord::as_ptr(&self)`

File: crates/flipperzero/src/dialogs/mod.rs
Comment (mogenson) [reply crates/flipperzero/src/dialogs/mod.rs:74]: Thanks for the tip. Will do!

File: crates/flipperzero/examples/storage.rs
Comment (mogenson) [reply crates/flipperzero/examples/storage.rs:28]: Good to know. Thanks.

File: crates/flipperzero/src/furi/string.rs
Comment (mogenson) [reply crates/flipperzero/src/furi/string.rs:90]: I was a bit unsure about this method because it's just a wrapper for the `as_ptr()` method of the core `NonNull` type. `NonNull<T>::as_ptr()` takes `self` and returns an `*mut T`.

Comment (str4d) [reply crates/flipperzero/src/furi/string.rs:90]: This method absolutely needs to take `&mut self`. AFAICT the only place this PR uses this method is `sys::dialog_file_browser_show`, which mutates the provided `sys::FuriString` to store the selected path. From the Rust perspective, this is mutating the `FuriString` and therefore requires exclusive access.

I also intentionally did not add this method to the public API, because I do not want it to exist there; it enables violating the safety requirements relied on by the `FuriString` struct. We should instead be crafting safe APIs that wrap the FFI methods taking `*mut sys::FuriString`, inside which we can call this in a crate-private setting where we know the safety requirements are satisfied.

Finally, we should name this `as_mut_ptr` to match equivalent APIs in the Rust stdlib ([`slice::as_mut_ptr`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.as_mut_ptr), [`Vec::as_mut_ptr`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.as_mut_ptr), [`str::as_mut_ptr`](https://doc.rust-lang.org/stable/std/primitive.str.html#method.as_mut_ptr)). I'd personally also prefer that this method be `unsafe`, though technically obtaining this kind of reference is safe Rust (it's just using said reference that is unsafe), so I'm not going to require that.
```suggestion
    pub(crate) fn as_mut_ptr(&mut self) -> *mut sys::FuriString {
```

File: crates/flipperzero/examples/storage.rs
Comment (dcoles) [reply crates/flipperzero/examples/storage.rs:28]: Oh, sorry. I don't know how I missed that. This is fine.

File: crates/flipperzero/src/furi/string.rs
Comment (dcoles) [reply crates/flipperzero/src/furi/string.rs:90]: @JarvisCraft As far as I'm aware methods like [`str::as_mut_ptr`](https://doc.rust-lang.org/std/primitive.str.html#method.as_mut_ptr) all need `&mut self` (or `self` so they consume the original object).

Comment (dcoles) [reply crates/flipperzero/src/furi/string.rs:90]: @str4d I think there's always going to be a need for folks to be able to get access to the underlying `*sys::FuriString` for any APIs that we haven't wrapped (e.g. maybe a third party Flipper Zero C library).

> I do not want it to exist there; it enables violating the safety requirements relied on by the `FuriString` struct

I'm trying to think of which safety requirements you could violate. My understanding was that `FuriString` was just a thin wrapper around `NonNull<sys::FuriString>`, so as long as you don't violate any standard ownership rules (e.g. calling `furi_string_free` on a string that you aren't the owner of), then it can be used safely.

I did see that our `FuriString` has the property "no nul bytes in the middle", but that might not be strictly necessary (if `m-string` allows it). While `CString` forbids an interior nul byte (I guess so the `into_raw`/`from_raw` pair works), `CStr` does not have that requirement.

Let me know if I'm missing something (which is well possible here).

Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string.rs:90]: In my opinion, adding `&mut` is not the right choice here.

It does not add to the safety since you [can](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ae5ff565df8bdd967ca844eeaefb1d78) safely convert `*const T` to `*mut T`. At the same time, this requirement would add a faulty feeling in safety to its user making him think that something is guaranteed by it (which is false).

Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string.rs:90]: From the other side, I like the crate-local access modifier so I generally like @str4d's suggestion.

Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string.rs]: Actually, you can propose this change to flipperdevices/flipperzero-firmware. I've done similar job with adding `const`-qualifiers and they were fine with merging them.

Comment (dcoles) [reply crates/flipperzero/src/furi/string.rs]: > Actually, `NonNull` is [`impl Copy`](https://doc.rust-lang.org/src/core/ptr/non_null.rs.html#701) so it does not effectively consume it.

Oh! I did not realize that was the case. OK. That makes more sense. Also I see `NonNull<T>` is defined as being the same as `*mut T`, so there's no need for a `mut self`, because `self` *is* a pointer-to-mut-T.

> In my opinion, adding `&mut` is not the right choice here.

I see this is analogous to `String::as_ptr(&self) -> *const u8` and `String::as_mut_ptr(&mut self) -> *mut u8`. Sure, you can freely cast back and forth between `*const` and `*mut` (since Rust won't let you actually _do_ anything with these pointers in safe code). However if you (the user) violate mutability rules (e.g. by creating a `mut*` to an immutable object and then mutatating it in `unsafe` code), then that is not Rust's fault. You, the user, are responsible for upholding Rust's mutability rules in `unsafe` code.

`const*` doesn't really mean all that much in Rust (or C). It's essentially type-checked documentation. About the only thing it does in Rust is [prevents you from performing mutable operations when you dereference it](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b95ce199ad8d90a13522a3c2a194dcb4).

Is it useless? No, it's still helpful to have the compiler act as a safety. The `furi_string_` methods are all const-correct and will stop you if you try to mutate something through a `const*`. If you take off the safety and point it at your foot, well you're in unsafe code, so neither Rust or C is going to stop you.

This would all be far less confusing if we could have done in-place/`#[repr(transparent)]` allocation of `FuriString` (flipperdevices/flipperzero-firmware#2461). Then it's super clear that you can't get a `*FuriString` that you can legally write through without having a `mut self` first.

> From the other side, I like the crate-local access modifier so I generally like @str4d's suggestion.

OK, I have no problem with keeping this `pub(crate)`.

Comment (dcoles) [reply crates/flipperzero/src/furi/string.rs]: > Maybe we can include a note in the comment doc that although this method needs a mutable reference, the starting file path will not be modified. It would be better if the `sys::dialog_show_file_browser()` C function took a `*const FuriString` for path instead of `*FuriString`.

This is one totally legitimate case for `*const T as *mut T`. It's fine to call `FuriString::as_ptr() -> *const sys::FuriString` and then cast it to a `*mut sys::FuriString` so long as you don't modify memory through this pointer.

Ultimately what matters (for correct compilation) is "does `sys::dialog_show_file_browser()` actually mutate the data behind the `path` pointer?", not whether the function declaration says `const *FuriString` or `*FuriString`.

**Comment by mogenson** [comment]
After inspecting the flipperzero-firmware code, `dialog_show_file_browser()` will not modify the preselected file `path` FuriString, but some common storage processing functions will replace `/ext`, `/app`, `/any` prefixes depending on the context. I think `pub(crate) fn as_mut_ptr(&mut self) -> *mut sys::FuriString` make sense, at least for the `show_file_browser()` usage. Is everyone in agreement?

**Comment by mogenson** [comment]
Also, is there anything I need to do with this PR to get the Clippy (MSRV) lints CI to succeed?

**Comment by mogenson** [comment]
@dcoles @JarvisCraft @str4d Are there any remaining issues or is this PR able to be merged?

**Review by JarvisCraft — APPROVED** [review]

**Comment by dcoles** [comment]
> Also, is there anything I need to do with this PR to get the Clippy (MSRV) lints CI to succeed?

@str4d Would you be able to take a look at this? I'm not quite sure why the beta is fine, but MSRV is throwing a `Unable to run @augu/clippy-action: HttpError: Resource not accessible by integration`.

**Comment by dcoles** [comment]
> @dcoles @JarvisCraft @str4d Are there any remaining issues or is this PR able to be merged?

Yes, let's get this merged. Thanks very much for your patience and for the PR!

**Comment by mogenson** [comment]
Cheers, thanks!

**Pull request merged**
